// watch-rx/dist/watch-rx.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.watchRx = watchRx;

var _rxjs = require('rxjs');

var _globResultFile = require('./glob-result-file');

var _chokidar = require('chokidar');

var _chokidar2 = _interopRequireDefault(_chokidar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function watchRx(pattern, options) {
  options = options || {};
  var basedir = options.cwd || process.cwd();

  return _rxjs.Observable.create(function (observer) {
    var isFinished = false;

    var watcher = _chokidar2.default.watch(pattern, options);
    var nextItem = function nextItem(event) {
      return function (name) {
        return observer.next(Object.assign(new _globResultFile.GlobResultFile(), {
          event: event,
          basedir: basedir,
          name: name.replace(/\\/g, '/')
        }));
      };
    };

    ['add', 'change', 'unlink', 'addDir', 'unlinkDir'].forEach(function (event) {
      watcher.on(event, nextItem(event));
    });

    watcher.on('error', function (err) {
      isFinished = true;
      observer.error(err);
      closeWatcher();
    });

    return function () {
      if (!isFinished) {
        closeWatcher();
      }
    };

    // Node doesn't exit after closing watchers
    // https://github.com/paulmillr/chokidar/issues/434
    function closeWatcher() {
      watcher.close();
    }
  });
}
